### PR TITLE
feat(calculate_hash): allow to customize hash input

### PR DESCRIPTION
**Pull request by sebffischer**
In `mlr3torch`, there are somewhat complicated "tensor-columns" that reference environments. 
Because the DataBackend hashes the whole data.table, this can lead to problems because of byte-code compilation (https://github.com/mlr-org/mlr3/blob/78ad4dd439bdc7ebfa2ca327eb6cb6e08c424e16/R/DataBackendDataTable.R#L159C24-L159C54). 

By merging this PR, it is possible to customize which information from a column is used to calculate the hash.

### CODE REVIEW

File: R/calculate_hash.R [hunk 1]
```diff
@@ -21,13 +11,44 @@
...
-    }
-  }), algo = "xxhash64")
+  digest(lapply(list(...), hash_input), algo = "xxhash64")
+}
+
+#' Hash Input
+#'
+#' Returns the part of an object to be used to calculate its hash.
+#'
+#' Methods exist for:
```
Comment (sebffischer) [inline R/calculate_hash.R:21]:
```suggestion
```

File: R/calculate_hash.R [hunk 2]
```diff
@@ -21,13 +11,43 @@
...
+#'
+#' @param x (any)\cr
+#'   Object for which to retrieve the hash input.
+#' @export
+hash_input = function(x) {
+  UseMethod("hash_input")
+}
+
+#' @describeIn hash_input
+#' The formals and the body are hashed separately.
```
Comment (sebffischer) [inline R/calculate_hash.R:29]:
```suggestion
#' The formals and the body are returned.
```

File: R/calculate_hash.R [hunk 3]
```diff
@@ -21,13 +11,43 @@
...
+#' This ensures that the bytecode or parent environment are not be included
```
Comment (sebffischer) [inline R/calculate_hash.R:30]:
```suggestion
#' This ensures that the bytecode or parent environment are not included.
```

File: R/calculate_hash.R [hunk 4]
```diff
@@ -21,13 +11,43 @@
...
+#' in the hash.
+#' @export
+hash_input.function = function(x) {
+  list(formals(x), as.character(body(x)))
+}
+
+#' @describeIn hash_input
+#' The data.table is converted to a regular list and.
```
Comment (sebffischer) [inline R/calculate_hash.R:38]:
```suggestion
#' The data.table is converted to a regular list and `hash_input()` is applied to all elements.
```

File: R/calculate_hash.R [hunk 5]
```diff
@@ -21,13 +11,43 @@
...
+#' This ensures that keys and indices are not included in the hash.
```
Comment (sebffischer) [inline R/calculate_hash.R:39]:
```suggestion
#' The conversion to a list ensures that keys and indices are not included in the hash.
```

File: R/calculate_hash.R [hunk 6]
```diff
@@ -21,13 +11,43 @@
...
+#' Then, [`hash_input`] is applied to each element of the list.
```
Comment (sebffischer) [inline R/calculate_hash.R:40]:
```suggestion
```

**Comment by sebffischer** [comment]
* [ ] revdepcheck

**Review by mb706 — COMMENTED** [review]
Maybe document the point of calculate_hash a bit more, otherwise ok with me.

File: R/calculate_hash.R [hunk 7]
```diff
@@ -1,17 +1,11 @@
...
-#' * If an object is a [data.table::data.table()], the data.table is converted to a
-#'   regular list. This ensures that keys and indices are not included in the hash.
-#'
-#' Note that this only applies to top level objects, these transformations are not done
-#' recursively.
+#' Calls [digest::digest()] using the 'xxhash64' algorithm after applying [`hash_input`] to each object.
+#' To customize the hashing behaviour, you can overwrite [`hash_input`] for specific classes.
+#' For `data.table` objects, [`hash_input`] is applied to all columns, so you can overwrite [`hash_input`] for
+#' columns of a specific class.
+#' Objects that don't have a specific method are hashes as is.
```
Comment (sebffischer) [inline R/calculate_hash.R:8]:
```suggestion
#' Objects that don't have a specific method are hashed as is.
```

**Pull request merged**